File: car-hunter-hamburg/frontend/src/api/history.ts
import { api } from "./client";
import type { Car } from "../types/car";

export interface PriceHistoryEntry {
  id: number;
  car_id: number;
  old_price: number;
  new_price: number;
  date: string;
  car_brand: string;
  car_model: string;
}

export function fetchPriceDrops(): Promise<PriceHistoryEntry[]> {
  return api.get<PriceHistoryEntry[]>("/api/history/price-drops");
}

export function fetchRemovedCars(): Promise<Car[]> {
  return api.get<Car[]>("/api/history/removed");
}

File: car-hunter-hamburg/backend/app/routers/history.py
"""
Endpoint /api/history — alimente la page "Historique" du frontend :
baisses de prix récentes et annonces devenues inactives (retirées par
le vendeur).
"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.car import Car
from app.models.price_history import PriceHistory
from app.schemas.price_history import PriceHistoryEntry
from app.schemas.car import CarOut

router = APIRouter()


@router.get("/price-drops", response_model=list[PriceHistoryEntry])
def price_drops(db: Session = Depends(get_db), limit: int = Query(30, le=100)):
    rows = (
        db.query(PriceHistory, Car.brand, Car.model)
        .join(Car, Car.id == PriceHistory.car_id)
        .order_by(PriceHistory.date.desc())
        .limit(limit)
        .all()
    )
    return [
        PriceHistoryEntry(
            id=ph.id, car_id=ph.car_id, old_price=ph.old_price, new_price=ph.new_price,
            date=ph.date, car_brand=brand, car_model=model,
        )
        for ph, brand, model in rows
    ]


@router.get("/removed", response_model=list[CarOut])
def recently_removed(db: Session = Depends(get_db), limit: int = Query(30, le=100)):
    """Annonces retirées récemment (probablement vendues ou dépubliées)."""
    return (
        db.query(Car)
        .filter(Car.is_active.is_(False))
        .order_by(Car.date_last_seen.desc())
        .limit(limit)
        .all()
    )
